libavformat/udp.c

```diff
@@ -494,6 +495,11 @@ static void *circular_buffer_task( void *_URLContext)
            in Single Unix. */
         pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &old_cancelstate);
         len = recv(s->udp_fd, s->tmp+4, sizeof(s->tmp)-4, 0);
```

Comment:
recv may not be interrupted correctly on all POSIX implementations. One expected behaviour would be that any interrupt (signal such as SIG_CANCEL_SIGNAL) will set errno to EINTR.

But BSD behaviour may just lead to an interrupt by our signal and restart the in progress function recv. In that case we may need to wait for any sort of interrupt or socket close that will let us continue the execution flow.

Some proposed fixed could include setting a timeout for recv until it continues with the regular execution. We could then also complete replace the cancel mechanism and use a volatile cancel flag.

We could also change the SA_RESTART flag to restart the interrupted system call (here recv). But that would require sigaction and have an influence on the signal action handler table. This table has a process-wide scope and would affect all threads within FFMpeg.